go/libkb/bug_3964_repairman.go

```diff
@@ -0,0 +1,232 @@
+
+	// Ignore any decoding errors
+	if tmpErr != nil {
+		b.G().Log.Warning("Problem reading previous bug 3964 repair time: %s", tmpErr)
+	}
+
+	if repairTime.IsZero() {
+		b.G().Log.Debug("| repair time is zero or wasn't set")
+		return false, nil
+	}
+	var fileTime time.Time
+	fileTime, err = StatSKBKeyringMTime(nun, b.G())
+	if err != nil {
+		return false, err
+	}
+	ss = !repairTime.Before(fileTime)
+	b.G().Log.Debug("| Checking repair-time (%s) v file-write-time (%s): shortCircuit=%v", repairTime, fileTime, ss)
+	return ss, nil
+}
+
+func (b *bug3964Repairman) fixLKSClientHalf(lctx LoginContext, lksec *LKSec, ppgen PassphraseGeneration) (err error) {
+	defer b.G().Trace("bug3964Repairman#fixLKSClientHalf", func() error { return err })()
+	var me *User
+	var encKey GenericKey
+	var ctext string
+
+	me, err = LoadMe(LoadUserArg{Contextified: NewContextified(b.G()), LoginContext: lctx})
+	if err != nil {
+		return err
+	}
+	encKey, err = me.GetDeviceSubkey()
+	if err != nil {
+		return err
+	}
+	// make client half recovery
+	kid := encKey.GetKID()
+	ctext, err = lksec.EncryptClientHalfRecovery(encKey)
+	if err != nil {
+		return err
+	}
+
+	_, err = b.G().API.Post(APIArg{
+		Endpoint:    "device/update_lks_client_half",
+		NeedSession: true,
+		Args: HTTPArgs{
+			"ppgen":           I{Val: int(ppgen)},
+			"kid":             S{Val: kid.String()},
+			"lks_client_half": S{Val: ctext},
+		},
+		SessionR: lctx.LocalSession(),
+	})
+
+	return err
+}
+
+// Run the engine
+func (b *bug3964Repairman) Run(lctx LoginContext, pps *PassphraseStream) (err error) {
+	defer b.G().Trace("bug3964Repairman#Run", func() error { return err })()
+
+	var lksec *LKSec
+	var ran bool
+	var dkm DeviceKeyMap
+	var ss bool
+	var serverHalfSet *LKSecServerHalfSet
+	nun := b.G().Env.GetUsername()
+
+	if b.G().TestOptions.NoBug3964Repair {
+		b.G().Log.Debug("| short circuit due to test options")
+		return nil
+	}
+
+	if pps == nil {
+		b.G().Log.Debug("| Can't run repairman without a passphrase stream")
+		return nil
+	}
+
+	if ss, err = b.computeShortCircuit(nun); err != nil {
+		return err
+	}
+
+	if ss {
+		// This logline is asserted in testing in bug_3964_repairman_test
+		b.G().Log.Debug("| Repairman already visited after file update; bailing out")
+		return nil
+	}
+
+	// This logline is asserted in testing in bug_3964_repairman_test
+	b.G().Log.Debug("| Repairman wasn't short-circuited")
+
+	lksec, err = pps.ToLKSec(b.G(), lctx.GetUID())
+	if err != nil {
+		return nil
+	}
+
+	if dkm, err = b.loadLKSecServerDetails(lctx, lksec); err != nil {
+		return err
+	}
+
+	if ran, serverHalfSet, err = b.attemptRepair(lctx, lksec, dkm); err != nil {
+		return err
+	}
+
+	if err != nil {
+		return err
+	}
+
+	b.G().Log.Debug("| SKB keyring repair completed; edits=%v", ran)
+
+	if !ran {
+		b.saveRepairmanVisit(nun)
+		return nil
+	}
+
+	if err := b.fixLKSClientHalf(lctx, lksec, pps.Generation()); err != nil {
```

Comment:
I think if this operation fails, we should retry (and therefore not save the visit)